### ISSUE TITLE
Annotation to allow mutating property in mutation-free context

### ISSUE THREAD

**Issue by AndrolGenhald** [vimeo/psalm#8437]
Sometimes it's desirable to have a property be modified in a mutation-free context when that modification wouldn't change the mutation-free status of the function: https://psalm.dev/r/ae13d40aee. It would be good to have an annotation for this, but we should also make sure the documentation warns users that using this annotation prevents safety checks, and it's the user's responsibility to ensure that the mutations on the property don't affect the observable behavior of the class.

See also doctrine/collections#322 where this came up with `ReadableCollection::contains()`.

**Comment by greg0ire** [comment]
I just noticed there is `@psalm-external-mutation-free`, does this match the feature request, or is that something else?

**Comment by AndrolGenhald** [comment]
That's probably why I thought we already had something for this, but that's actually used to annotate methods that are allowed to modify properties on `$this` but nothing else.

**Comment by greg0ire** [comment]
I believe that would be appropriate for `contains()`, but using it does not seem to help:

```
ERROR: ImpureMethodCall - lib/Doctrine/Common/Collections/AbstractLazyCollection.php:70:35 - Cannot call a possibly-mutating method Doctrine\Common\Collections\Collection::contains from a mutation-free context (see https://psalm.dev/203)
        return $this->collection->contains($element);


ERROR: InvalidTemplateParam - lib/Doctrine/Common/Collections/ReadableCollection.php:24:21 - Template param T of Doctrine\Common\Collections\ReadableCollection is marked covariant and cannot be used here (see https://psalm.dev/183)
     * @psalm-param T $element


------------------------------
2 errors found
------------------------------
3 other issues found.
You can display them with --show-info=true
------------------------------
```

**Comment by AndrolGenhald** [comment]
Right, because since it's allowed to modify properties on `$this` it might cause the template type to change. [This](https://psalm.dev/r/f3cfa359f0) is fine, [this](https://psalm.dev/r/1ff219ae6d) is not. Tbh I'm not really sure what the use case of `@psalm-external-mutation-free` is, I've never used it or seen it used.

**Comment by greg0ire** [comment]
Ok so we do indeed need something more precise, where there might be an internal mutation as long as it does not affect the template item type.

**Comment by danog** [comment]
Precisely what I need for the `checked` field of Atomic in my immutable refactoring :)
